### PR TITLE
Device dmatrix

### PR DESCRIPTION
**Pull request by RAMitchell**
This is a proof of concept for a DMatrix class that only contains an EllPack page and may be created efficiently from cupy/cudf inputs. See #5143 for roadmap.

Issues to be resolved:
- The DeviceDMatrix currently has no way to get the parameter `max_bin` so is fixed on 256 bins. 
- I directly exposed the DMatrix class in python. Not sure if this is necessary or the best way to do things.
- 2 new C API functions are added. Again not sure if this is the best approach or we should pass parameters into existing functions to select DMatrix type.

### CODE REVIEW

**Comment by trivialfis** [comment]
I prefer the name `HistogramDMatrix`/`QuantileDMatrix`, as it's created for histogram algorithm.

**Comment by RAMitchell** [comment]
Should we separate implementations for device and host? That would be `DeviceQuantileDMatrix`

**Comment by trivialfis** [comment]
I think a general interface is better.

**Comment by RAMitchell** [comment]
@trivialfis Tests are passing and I have added support for max_bin as a DMatrix parameter. I named the python class DeviceQuantileDMatrix, after some thought I don't want to hedge the CPU version of this class into the same python wrapper. I think it's better to have a separate class. I prefer being explicit about what it is upfront instead of doing automatic magic underneath to switch between functionality.

File: python-package/xgboost/core.py [hunk 1]
```diff
@@ -1001,6 +1001,69 @@ def feature_types(self, feature_types):
         self._feature_types = feature_types
 
 
+class DeviceQuantileDMatrix(DMatrix):
+    """Device memory Data Matrix used in XGBoost.
+
+    DMatrix is a internal data structure that used by XGBoost
```
Comment (trivialfis) [inline python-package/xgboost/core.py]: Do we need to mention `gpu_hist` explicitly ?

File: python-package/xgboost/core.py [hunk 2]
```diff
@@ -1001,6 +1001,69 @@ def feature_types(self, feature_types):
...
+    def __init__(self, data, label=None, weight=None, base_margin=None,
+                 missing=None,
+                 silent=False,
+                 feature_names=None,
+                 feature_types=None,
+                 nthread=None, max_bin=256):
+        self.max_bin = max_bin
+        if not (hasattr(data, "__cuda_array_interface__") or (
+                CUDF_INSTALLED and isinstance(data, CUDF_DataFrame))):
+            print(type(data))
```
Comment (trivialfis) [inline python-package/xgboost/core.py]: remove

File: src/data/device_dmatrix.cu [hunk 3]
```diff
@@ -0,0 +1,251 @@
...
+#include "../common/math.h"
+#include "adapter.h"
+#include "device_adapter.cuh"
+#include "ellpack_page.cuh"
+#include "device_dmatrix.h"
+
+namespace xgboost {
+namespace data {
+
+struct IsValidFunctor : public thrust::unary_function<Entry, bool> {
```
Comment (trivialfis) [inline src/data/device_dmatrix.cu]: Is this copied from `simple_dmatrix.cu`?

File: src/data/device_dmatrix.cu [hunk 4]
```diff
@@ -0,0 +1,251 @@
...
+                                         device_accessor.NullValue(), idx);
+    }
+  });
+}
+
+// Does not currently support metainfo as no on-device data source contains this
+// Current implementation assumes a single batch. More batches can
+// be supported in future. Does not currently support inferring row/column size
+template <typename AdapterT>
+DeviceDMatrix::DeviceDMatrix(AdapterT* adapter, float missing, int nthread, int max_bin) {
```
Comment (trivialfis) [inline src/data/device_dmatrix.cu]: Is the weight used?

File: python-package/xgboost/core.py (hunk 1 above)
Comment (RAMitchell) [reply python-package/xgboost/core.py]: Done, added a little more documentation.

File: src/data/device_dmatrix.cu (hunk 3 above)
Comment (RAMitchell) [reply src/data/device_dmatrix.cu]: Yes it's the same but I don't see a convenient place to put it.

File: src/data/device_dmatrix.cu (hunk 4 above)
Comment (RAMitchell) [reply src/data/device_dmatrix.cu]: No, I added some mention of this in the python docstring.

Comment (trivialfis) [reply src/data/device_dmatrix.cu]: @RAMitchell We might need to be careful around this as changing the C api in the future is not easy.

File: python-package/xgboost/core.py [hunk 5]
```diff
@@ -1001,6 +1001,71 @@ def feature_types(self, feature_types):
...
+                         missing=missing,
+                         silent=silent,
+                         feature_names=feature_names,
+                         feature_types=feature_types,
+                         nthread=nthread)
+
+    def _init_from_array_interface_columns(self, df, missing, nthread):
+        """Initialize DMatrix from columnar memory format."""
+        interfaces = []
+        for col in df:
```
Comment (trivialfis) [inline python-package/xgboost/core.py]: Could you extract this as a function.  I believe there's a same portion of code in DMatrix.

File: src/data/device_dmatrix.cu (hunk 3 above)
Comment (trivialfis) [reply src/data/device_dmatrix.cu]: > Yes it's the same but I don't see a convenient place to put it.

`adapter.cuh`?

**Comment by trivialfis** [comment]
Please rebase onto master for resolving failure on travis.

**Pull request merged**